arch/arm: introduce the pre-stack/RAM init hook

nvlsianpu opened the pull request:
Introduce hook for customize reset.S code even before stack is initialized or RAM is accessed. Hook can be enabled using CONFIG_SOC_EARLY_RESET_HOOK=y.
Hook implementation is by soc_start_hook() function which should be provided by custom code.

For my case this hook can be use for implementing S2RAM resume operation bridge by the zephyr-rtos based bootloader: for such case the startu code is redirected to application startup code via application's reset vector call.
In the hook proram does recognition of whether this is the regular CPU boot or the S2RAM resume boot. For regular boot the hook returns to reset.S code flow. For S2RAM resume boot the hooks performs jump to well-known (for bootloader) application reset vector.

``` TXT
              resume/reboot                                               reboot==T
[S2RAM sleep]-------------->[ boot (reset.S) ]---->[soc_early_reset_hook]---------->[ continue regular boot ]
                                                          |  
                                                          |      resume==T
                                                          '----------------------> [ s2ram resume ]
```

This PR supersede #96290 as it is more generic an SoC independent.

nvlsianpu committed:
arch/arm: introduce the pre-stack/RAM init hook

Introduce hook for customize reset.S code even before stack is
initialized or RAM is accessed. Hook can be enabled using
CONFIG_SOC_EARLY_RESET_HOOK=y.
Hook implementation is by soc_early_reset_hook() function which should
be provided by custom code.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

bjarki-andreasen commented:
This hook is also useful for applying trims/factory calibrations to RAM blocks before they are accessed which is the case for some Nordic SoCs.

wearyzen commented on include/zephyr/platform/hooks.h:
```diff
@@ -19,6 +19,15 @@
+ * SoC-specific initialization. Its execution takes place even before
+ * the stack is initialized or RAM is accessed.
+ */
+void soc_pre_ram_hook(void);
```
I would suggest to change it to `soc_resume_hook` or `soc_pm_resume_hook` or `soc_early_resume_hook` because you want this to be call immediately after resume without the arch doing anything. The "pre-ram" part suggests there is still something that arch's do and it could be setting SP which you are not expecting in this hook.

nvlsianpu replied:
Indeed I need this for s2ram resume.

Noticed @bjarki-andreasen comment about other possible objective:
>This hook is also useful for applying trims/factory calibrations to RAM blocks before they are accessed which is the case for some Nordic SoCs.

Probably this hook can be used to many peculiar things we don't know yet.
I'm open for som better name.

wearyzen replied:
I would vote for `soc_resume_hook` to keep things simple and the description would cover that this is called before arch's own resume logic.

nashif replied:
yeah, the more generic, the better, you can add the pre-stack/ram usage as an example in the kconfig docs.

mathieuchopstm commented on kernel/Kconfig.init:
```diff
@@ -6,6 +6,15 @@
 
 menu "SoC and Board Hooks"
 
+config SOC_PRE_RAM_HOOK
+	bool "Run early SoC pre-RAM reset hook"
+	help
+	  Run an early SoC reset hook, before RAM get initialize.
+
+	  A custom hook soc_pre_ram_hook() is executed at the beginning of the
+	  startup code (__start), even before any stack is initiated
+	  or RAM is accessed. soc_pre_ram_hook() must be implemented by the SoC.
```
Outside PR scope, but then I would change `SOC_RESET_HOOK` to be described as `called after "soc_pre_ram_hook", has a valid stack [with size XXX if that may be portable]`.

And I'd reword all symbols to use the `Run SoC-specific XXX hook` formulation. Though again there's some clarification here to be made about who can implement what hook: maybe some should be reserved for SoC-specific code only, whereas some are more likely to be used by application-specific code.

nvlsianpu replied:
used your suggestion on Kconfig help.

bjarki-andreasen replied:
So, the hook could be used for many things, its not just there to do resume from S2RAM, that's just one thing it can be used for:
```
/* this would probably be ASM */
void soc_early_init_hook(void)
{
        soc_trim_ram_blocks();

#if SOC_DIRECT_JUMP_TO_APP
        /* jump to app */
#endif

        ...
}
```
The hook is not "meant" for resume, its just a place this particular way of implementing S2RAM can utilize.

wearyzen replied:
> The whole reason for these PRs was that one wanted `arch_pm_s2ram_resume` called on a different stack so it sounded like a reasonable required. I couldn't think of a reason to run code in a hook _before `arch_pm_s2ram_resume`_[1] other than "I want to override S2RAM resume in an application-specific manner - e.g., I'm a bootloader and will hand over control to loaded image", which would require a specific stack.
> 
with `soc_resume_hook` there won't be a need to return and call `arch_pm_s2ram_resume`. Ref. https://github.com/zephyrproject-rtos/zephyr/pull/96290#issuecomment-3333319836

The way I see it is:
- In a reset case, `soc_resume_hook` returns and the architecture's reset handler continues. The arch reset handler doesn't care what was msp/psp or what was on it since it is going to clear it and initialize it.
- In a resume case where soc wants to resume on their own and don't need `arch_pm_s2ram_resume`, `soc_resume_hook` continues and doesn't return. In this case, like explained in the link above soc can change sp, push stuff on it and do anything with it.
- In resume case where soc wants to use `arch_pm_s2ram_resume` but also wants to do stuff before it, they can still change sp, push stuff on stack etc. because maybe that is what they want to do and then after returning from `soc_resume_hook` arch's code continues to call `arch_pm_s2ram_resume`. If I look at arm reset handler, it is switching to use interrupt stack after `soc_resume_hook` so again it doesn't matter what they did inside `soc_resume_hook` 🤷 .

When I lay out these 3 cases I don't think we need to put the stack or other restrictions on the hook.

bjarki-andreasen replied:
In that case, `soc_pre_reset_hook` (or early) sounds like it describes it better than `soc_resume_hook` which only namely matches the one case where it does not return?

bjarki-andreasen replied:
to be clear though, we have not applied the trim yet, and nothing has broken yet so it seems the defaults are stable enough, but according to specs trims should be applied first.

mathieuchopstm replied:
My two cents: we should keep the "no stack" restriction and, *if* a usecase where stack is needed ***and*** `soc_reset_hook` isn't sufficient appears, the decision can be reconsidered. *But I won't block on this point.*

For the name, `soc_early_reset_hook` is fine by me.

BTW, I would update the documentation for `soc_reset_hook` to indicate it runs after the architecture-specific resume path - and thus isn't executed on warm boot path.

mathieuchopstm replied:
Right - `s2ram_mark_check` could require restrictions to be imposed on it which would be painful (i.e., "no stack usage => must be ASM" - which FTR Nordic folks were willing to do in the first place, and is done on STM32WB0 too now)

bjarki-andreasen reviewed (dismissed):
I think this looks good

nvlsianpu commented on kernel/Kconfig.init:
```diff
@@ -6,6 +6,27 @@
+	  Run a SoC-specific hook early in the reset/startup code (__start).
+	  A custom hook soc_pre_ram_hook() will be executed at the beginning
+	  of the architecture-specific startup code, after hardware has been
+	  configured as required by CONFIG_INIT_ARCH_HW_AT_BOOT.
```
used

wearyzen commented on include/zephyr/platform/hooks.h:
```diff
@@ -19,6 +19,20 @@
+ * of the reset vector.
+ *
+ * This hook is implemented by the SoC and can be used to perform any
+ * SoC-specific initialization. Its execution might takes place even before
+ * the stack is initialized or data RAM is accessed.
+ * Refer to relevant architecture zephyr-rtos startup implementation for more details.
```
nit:

```suggestion
 * SoC-specific initialization. Refer to :kconfig:option:`SOC_EARLY_RESET_HOOK` and relevant architecture zephyr-rtos startup implementation for more details.
```

mathieuchopstm commented on kernel/Kconfig.init:
```diff
@@ -6,6 +6,30 @@
+	  special logic to resume execution in some scenarios (for example, when
+	  a bootloader is used).
+
+	  The hook is called immediately after reset so use stack wisely if
+	  the hook intends to return to continue with arch's resume/reset activity.
```
The two sentences are contradictory: either you can
> use stack wisely

or

> the stack pointer register should be considered as not initialized [...and you are] NOT allowed to push any data using this stack pointer value

Since it seems we agreed upon "no stack usage allowed" - and to avoid repetitions - I propose to drop the paragraph entirely:
```suggestion
```

mathieuchopstm commented on kernel/Kconfig.init:
```diff
@@ -6,6 +6,30 @@
+
+	  The hook is called immediately after reset so use stack wisely if
+	  the hook intends to return to continue with arch's resume/reset activity.
+	  The stack pointer register should be considered as not initialized by the caller of
+	  this hook. In particular, this means that the hook is NOT allowed to
+	  "push" any data using this stack pointer value. However, the hook may use
+	  an implementation-specific area as stack if desired; in such case,
+	  the original value of the stack pointer needs not to be "restored"
+	  before returning control to the hook's caller.
```
Nit: align lines to similar length (+ small reword):

```suggestion
	  The stack pointer register should be considered as not initialized upon
	  call to this hook. In particular, this means that the hook is NOT allowed
	  to "push" any data using this stack pointer value. However, the hook may
	  use an implementation-specific area as stack if desired; in such case,
	  the original value of the stack pointer needs not to be "restored" before
	  returning control to the hook's caller.
```

mathieuchopstm commented on kernel/Kconfig.init:
```diff
@@ -6,6 +6,30 @@
+	  Additional constraints may be imposed on the hook by the architecture.
+
 config SOC_RESET_HOOK
 	bool "Run early SoC reset hook"
```
Nit: I'd update `SOC_RESET_HOOK` while at it
```diff
config SOC_RESET_HOOK
-	bool "Run early SoC reset hook"
+	bool "Run SoC-specific reset hook"
	help
-	  Run an early SoC reset hook.
+	  Run a SoC-specific hook in the reset/startup code (__start).

-	  A custom hook soc_reset_hook() is executed at the beginning of the
-	  startup code (__start). soc_reset_hook() must be implemented by the SoC.
+	  A custom hook soc_reset_hook() will be executed near the beginning
+	  of the architecture-specific startup code, after hardware has been
+	  configured (as required by CONFIG_INIT_ARCH_HW_AT_BOOT), a stack 
+	  pointer has been loaded and the architecture's own resume logic
+	  has executed (if CONFIG_PM_S2RAM is enabled). Because this hook
+	  runs after the resume logic, it is not called when the system
+	  resumes from a suspend-to-RAM power state.
```

mathieuchopstm reviewed (commented):
Also:

nvlsianpu committed:
kernel/Kconfig.init: update description of SOC_RESET_HOOK

Updated description on conditions and assumptions in which
the soc_reset_hook is executed.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

andyross reviewed (approved):
No reason not to +1 as this is harmless.  But I think the real issue here is that there's tension between the "architecture reset vector", which is owned by arch/arm, and "platform reset", which is squishier and demands board-specific code.

On other architectures, the responsibility can be inverted: the SOC code takes the initial reset and calls into the OS when ready, using the arch layer as a "toolkit" for stuff that needs to be initialized but otherwise controlling the flow.  The various adsp platforms look like this (because they have complicated initialization that actually starts across a PCI bus in a Linux driver).  It might be something to consider more generally.
